File: modules/core/python/tests/test_normalize_meshes.py
#
# Copyright 2022 Adobe. All rights reserved.
# This file is licensed to you under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License. You may obtain a copy
# of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under
# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
# OF ANY KIND, either express or implied. See the License for the specific language
# governing permissions and limitations under the License.
#
import lagrange

import numpy as np
from numpy.linalg import norm
import math
import pytest


class TestNormalizeMeshes:
    def test_empty_mesh(self):
        mesh = lagrange.SurfaceMesh()
        lagrange.normalize_mesh(mesh)
        assert mesh.num_vertices == 0

    def test_single_point(self):
        mesh = lagrange.SurfaceMesh()
        mesh.add_vertex([0, 0, 0])
        lagrange.normalize_mesh(mesh)
        assert mesh.num_vertices == 1

    def test_triangle(self):
        mesh = lagrange.SurfaceMesh()
        mesh.vertices = np.array([[0, 0, 0], [10, 0, 0], [0, 10, 0]])
        lagrange.normalize_mesh(mesh)
        bbox_max = np.amax(mesh.vertices, axis=0)
        bbox_min = np.amin(mesh.vertices, axis=0)
        bbox_center = (bbox_max + bbox_min) / 2
        assert bbox_center == pytest.approx(np.zeros(3))
        assert norm(bbox_max - bbox_min) == pytest.approx(2)

    def test_small_triangle(self):
        mesh = lagrange.SurfaceMesh()
        mesh.vertices = np.array([[0, 0, 0], [1e-3, 0, 0], [0, 1e-3, 0]])
        lagrange.normalize_mesh(mesh)
        bbox_max = np.amax(mesh.vertices, axis=0)
        bbox_min = np.amin(mesh.vertices, axis=0)
        bbox_center = (bbox_max + bbox_min) / 2
        assert bbox_center == pytest.approx(np.zeros(3))
        assert norm(bbox_max - bbox_min) == pytest.approx(2)

    def test_two_meshes(self):
        mesh1 = lagrange.SurfaceMesh()
        mesh1.vertices = np.array([[0, 0, 0], [1, 0, 0], [0, 1, 0]])

        mesh2 = lagrange.SurfaceMesh()
        mesh2.vertices = np.array(
            [
                [0, 0, 10],
                [1, 0, 10],
                [0, 1, 10],
            ]
        )

        lagrange.normalize_meshes([mesh1, mesh2])

        vertices = np.vstack([mesh1.vertices, mesh2.vertices])
        bbox_max = np.amax(vertices, axis=0)
        bbox_min = np.amin(vertices, axis=0)
        bbox_center = (bbox_max + bbox_min) / 2
        assert bbox_center == pytest.approx(np.zeros(3))
        assert norm(bbox_max - bbox_min) == pytest.approx(2)

File: modules/core/tests/test_normalize_meshes.cpp
/*
 * Copyright 2022 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
#include <lagrange/testing/common.h>

#include <lagrange/normalize_meshes.h>

TEST_CASE("normalize_meshes", "[surface][utilities]")
{
    using namespace lagrange;
    using Scalar = double;
    using Index = uint32_t;

    SECTION("Single quad")
    {
        lagrange::SurfaceMesh<Scalar, Index> mesh;
        mesh.add_vertex({0, 0, 0});
        mesh.add_vertex({2, 0, 0});
        mesh.add_vertex({2, 2, 0});
        mesh.add_vertex({0, 2, 0});
        mesh.add_quad(0, 1, 2, 3);

        normalize_mesh(mesh);
        for (Index vi = 0; vi < 4; vi++) {
            auto v = mesh.get_position(vi);
            REQUIRE(v[0] >= -1);
            REQUIRE(v[0] <= 1);
            REQUIRE(v[1] >= -1);
            REQUIRE(v[1] <= 1);
            REQUIRE(v[2] >= -1);
            REQUIRE(v[2] <= 1);
        }
    }

    SECTION("Single quad with transform")
    {
        lagrange::SurfaceMesh<Scalar, Index> mesh;
        mesh.add_vertex({0, 0, 1});
        mesh.add_vertex({2, 0, 1});
        mesh.add_vertex({2, 2, 1});
        mesh.add_vertex({0, 2, 1});
        mesh.add_quad(0, 1, 2, 3);

        const auto transform = normalize_mesh_with_transform<3>(mesh);
        for (Index vi = 0; vi < 4; vi++) {
            auto v = mesh.get_position(vi);
            REQUIRE(v[0] >= -1);
            REQUIRE(v[0] <= 1);
            REQUIRE(v[1] >= -1);
            REQUIRE(v[1] <= 1);
            REQUIRE(v[2] >= -1);
            REQUIRE(v[2] <= 1);
        }

        auto transform_ = Eigen::Transform<Scalar, 3, Eigen::Affine>::Identity();
        transform_.translation().array() = 1.0;
        transform_.linear().diagonal().array() = sqrt(2.0);

        const auto transform_error = (transform.matrix() - transform_.matrix()).array().abs().sum();
        REQUIRE(transform_error < 1e-7);
    }

    SECTION("One quad and one triangle")
    {
        lagrange::SurfaceMesh<Scalar, Index> mesh;
        mesh.add_vertex({0, 0, 0});
        mesh.add_vertex({2, 0, 0});
        mesh.add_vertex({2, 2, 0});
        mesh.add_vertex({0, 2, 0});
        mesh.add_vertex({3, 1, 0});
        mesh.add_quad(0, 1, 2, 3);
        mesh.add_triangle(4, 2, 3);

        normalize_mesh(mesh);
        for (Index vi = 0; vi < 4; vi++) {
            auto v = mesh.get_position(vi);
            REQUIRE(v[0] >= -1);
            REQUIRE(v[0] <= 1);
            REQUIRE(v[1] >= -1);
            REQUIRE(v[1] <= 1);
            REQUIRE(v[2] >= -1);
            REQUIRE(v[2] <= 1);
        }
    }
}
